inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/NewClassDiscoveryService.java

```diff
@@ -0,0 +1,102 @@
+package rocks.inspectit.oce.core.instrumentation;
+
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import rocks.inspectit.oce.core.config.InspectitEnvironment;
+import rocks.inspectit.oce.core.utils.StopWatch;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import java.lang.instrument.Instrumentation;
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+@Service
+@Slf4j
+public class NewClassDiscoveryService implements IClassDefinitionListener {
+
+    @Autowired
+    private InspectitEnvironment env;
+
+    @Autowired
+    private Instrumentation instrumentation;
+
+    @Autowired
+    private ScheduledExecutorService exec;
+
+    @Autowired
+    private List<IClassDiscoveryListener> listeners;
+
+    private Set<Class<?>> knownClasses = Collections.newSetFromMap(new WeakHashMap<>());
+
+    private volatile boolean isShuttingDown = false;
+
+    /**
+     * Future for the scheduled {@link #updateCheckTask} to make it cancelable.
+     */
+    private Future<?> updateCheckFuture;
+
+    /**
+     * Stores the timestamp for the last time a class was defined.
+     */
+    private AtomicLong lastNewClassDefinitionTimestamp = new AtomicLong();
```

Comment:
I think you don't need an ``AtomicLong`` here. A ``private volatile long`` would be sufficient because you just direclty set/read the variable which are atomic operations by default (but you need the volatile on long).